Include definitions from GitHub repo

phackstock opened the pull request:
First version of the external repo fetching feature.

### How to use
In the `definitions` folder of a project place a file called definitions with the following content:

```yaml
region:
  repository: https://github.com/IAMconsortium/common-definitions.git/
  # hash: optionally specify a commit hash of the repository
  # release: optionally specify a release of the repository
```
the `hash` and `release` attributes are both optional. 
If neither is given, the last version of the main branch is used. 
They are also mutually exclusive. If both are given an error is raised.

### Implementation details

The repo fetching logic is implemented as follows:
1. If the repo is not present, clone it
2. Perform a hard reset to get rid of any changes
3. Check out the desired revision

danielhuppmann commented on tests/data/general-config-definitions/config.yaml:
```diff
@@ -1,3 +1,3 @@
 region:
```
We don't want to clone the same repo for variables, regions, etc.
```suggestion
repository:
  common-defintions:
    git: git@github.com:IAMconsortium/common-definitions.git
    hash:
    release:
region:
    repository:
```

phackstock replied:
> We don't want to clone the same repo for variables, regions, etc.

Do you mean cloning the same repo multiple times? 
If so, this wouldn't happen anyway since the code checks if the repo is present. If it is it is just updated if need be. 

Although I do have to say it might be nice to not have to define the repo url multiple times.

> Plus the "dictionary" of a repo-name to the git-attributes can help to support multiple external repos in the future.

This should already work.

```yaml
region:
  repository: https://github.com/IAMconsortium/common-definitions.git/
  # hash: optionally specify a commit hash of the repository
  # release: optionally specify a release of the repository
variable:
  # use a different repo for the variables
  repository: https://github.com/IAMconsortium/common-definitions.git/
  # hash: optionally specify a commit hash of the repository
  # release: optionally specify a release of the repository
```
unless you mean multiple external repos for a single dimension. In this case the code would need to be modified but I'd say we cross that bridge when we get there.

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -1,12 +1,44 @@
+    def repository_name(self):
+        return self.repository.split("/")[-2].split(".")[0]
+
+    def fetch_repo(self, to_path="./common-definitions"):
```
Why is the *to_path* hard-coded?

phackstock replied:
That was a left over from initial development, it's not removed.

danielhuppmann commented:
The implementation looks good, thanks! But I still think that the configuration would be more elegant and future-proof if we do:

```yaml
repository:
  common-definitions:
    git: git@github.com:IAMconsortium/common-definitions.git
    hash:  # optional
    release:  # optional
region:
    repository: common-definitions
    path:  # optional, defaults to `definitions / regions`
variable:
    repository: common-definitions
    path:  # optional, defaults to `definitions / variable`
```

This would make the *repository_name* explicit and customizable.

And the `fetch_repo()` should happen when initializing the **DataStructureConfig** (if a *repository* is given), so that the **CodeListConfig** only has to know which path (`"repository" / <repository_name> / <path>`) to parse in addition to its "own" (local) folder.

phackstock commented:
> The implementation looks good, thanks! But I still think that the configuration would be more elegant and future-proof if we do:

Agreed, I'll change it that way.

phackstock commented:
# Update

@danielhuppmann I have updated `DataStructureConfig` and all associated classes according to your suggestions. 
The configuration file looks now like this:
```yaml
repository:
  common-definitions:
    url: https://github.com/IAMconsortium/common-definitions
region:
  repository: common-definitions
  country: true
variable:
  repository: common-definitions
  repository_definition_path: definitions / variable
```

I changed your `git` attribute to `url` but happy to change it back. The repository url needs to needs to end with _.git/_ otherwise the _gitpython_ library fails to fetch the repository. 
I've also changed the `path` attribute in the dimensions, _region_ and _variable_, in this example, to `repository_definition_path`. I understand that it's a bit clunky and will probably be changed but I was confused by just the name `path`. Reason being is that it could also refer to the path of the local repositories codelist folder. Happy to suggestions to rename. Or even leave out for now as I think we can enforce the structure of an external repository to be "repo/definitions/_dimension_".

I've also, according to your suggestion moved the repository fetching into the constructor of `DataStructureConfiguration`, much cleaner this way. Now all repos are automatically fetched upon initialization.

In addition I've extended the use of external repos from just region to any dimension. 
The test that I've modified to cover this also tests for the case that there is no local codelist folder. In case of the test, there's no variable folder. Works perfectly.

danielhuppmann commented on tests/test_definition.py:
```diff
@@ -43,19 +44,26 @@ def test_empty_codelist_raises():
+        # added via general-config definitions in addition to pycountry.countries
+        assert "Kosovo" in obs.region
+
+        assert "Primary Energy" in obs.variable
```
```suggestion
        # imported from https://github.com/IAMconsortium/common-definitions repo
        assert "Primary Energy" in obs.variable
```

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -1,18 +1,69 @@
-    repository: Optional[Path]
+    dimension: str
+    repository: Optional[str]
+    repository_definition_path: Optional[str]
```
This does not have the "definitions" path, but the dimension within a definitions folder...

Also, should this be a `Path` object?
```suggestion
    repository_dimension_path: Optional[Path]
```

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -1,18 +1,69 @@
+    url: str
+    hash: Optional[str]
+    release: Optional[str]
+    path: Optional[Path]
```
Suggest to rename to clarify that this is the local (cloned) path...
```suggestion
    local_path: Optional[Path]  # defined via the `repository` name in the configuration
```

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -1,18 +1,69 @@
+    @validator("path")
+    def check_path_empty(cls, v):
+        if v is not None:
+            raise ValueError("path must not be set as part of the config")
```
```suggestion
            raise ValueError("The `path` must not be set as part of the config.")
```

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -23,7 +74,32 @@ class DataStructureConfig(BaseModel):
+                raise ValueError(
+                    (
+                        f"Unknown repository {values.get(dimension).repository} in"
+                        " region.repository."
```
```suggestion
                        f" {dimension}.repository."
```

danielhuppmann commented on nomenclature/config.py:
```diff
@@ -23,7 +74,32 @@ class DataStructureConfig(BaseModel):
+            ):
+                raise ValueError(
+                    (
+                        f"Unknown repository {values.get(dimension).repository} in"
```
```suggestion
                        f"Unknown repository '{values.get(dimension).repository}' in"
```

danielhuppmann reviewed (approved):
Very nice! A few suggestions for alternative keyword names, but good to be merged either way...
